### ISSUE TITLE
failed to open stream: No such file or directory

### ISSUE THREAD

**Issue by farazmunj** [kohana/cache#84]
on line 470 I get so many RuntimeException (s) there should be a check around it if file exists or not before it? 

RuntimeException [ 0 ]: SplFileInfo::openFile(.../cache/3a/3ae1f572b12efae209db8e0db92985eb10c66631.cache): failed to open stream: No such file or directory
MODPATH/cache/classes/Kohana/Cache/File.php [ 470 ]
465      */
466     protected function _is_expired(SplFileInfo $file)
467     {
468         // Open the file and parse data
469         $created = $file->getMTime();
470         $data = $file->openFile("r");
471         $lifetime = (int) $data->fgets();
472 
473         // If we're at the EOF at this point, corrupted!
474         if ($data->eof())
475         {
MODPATH/cache/classes/Kohana/Cache/File.php [ 470 ] » SplFileInfo->openFile(arguments)
0  
string(1) "r"
465      */
466     protected function _is_expired(SplFileInfo $file)
467     {
468         // Open the file and parse data
469         $created = $file->getMTime();
470         $data = $file->openFile("r");
471         $lifetime = (int) $data->fgets();
472 
473         // If we're at the EOF at this point, corrupted!
474         if ($data->eof())
475         {
MODPATH/cache/classes/Kohana/Cache/File.php [ 144 ] » Kohana_Cache_File->_is_expired(arguments)
0  
object SplFileInfo(0) {
}
139                 return $default;
140             }
141             else
142             {
143                 // Test the expiry
144                 if ($this->_is_expired($file))
145                 {
146                     // Delete the file
147                     $this->_delete_file($file, FALSE, TRUE);
148                     return $default;
149                 }
MODPATH/redback/classes/Cache/File.php [ 15 ] » Kohana_Cache_File->get(arguments)
0  
string(13) "widget_170_en"
1  
NULL
10         if (Kohana::$profiling === TRUE and class_exists('Profiler', FALSE)) {
11             // Start a new benchmark
12             $benchmark = Profiler::start('Cache get', __FUNCTION__ . ' ' . $id);
13         }
14         if (Kohana::$caching) {
15             $return = parent::get($id, $default);
16         }
17         if (isset($benchmark)) {
18             // Stop the benchmark
19             Profiler::stop($benchmark);
20         }

**Comment by acoulton** [comment]
There is already a test for file existence at [line 136](https://github.com/kohana/cache/blob/3.3/master/classes/Kohana/Cache/File.php#L136) so if `is_expired` is being called the file must have existed a few milliseconds earlier.

Most likely, I guess, you're seeing a race condition caused by the combination of a lot of concurrent requests and either using short cache TTLs or manually deleting cache entries. One request is deleting the file just as a new request arrives, so the file exists at line 136 but is gone by line 470.

Adding a second `isFile` check just before `openFile` would reduce the window for this and so might reduce the frequency, but it's unlikely to resolve it completely.

Possibly we could fix this by making the cache library catch and ignore these errors - but I'm tempted to say that you may be using the file driver outside the bounds of recommended usage, and you'd be better to switch to a more concurrency-safe cache backend.

**Comment by farazmunj** [comment]
Good point about combination of lot of concurrent requests.
I made some change in code yesterday and wrapped it in if($return = Cache::instance(CACHE_FILE)->get('id') ) and it reduced the number of errors since that, Strange.
I will keep eye on this issue and see if it bubble up again. its kind of same issue i face with session on that server it generate so many Session_Exception (s) as well.

**Comment by farazmunj** [comment]
Hi,
there is significant reduction in errors, Its still generating this error 2-5 times a day. I think its good idea to to check if file exists just before opening it.
So my final solution is extend Kohana_Cache_File 
and overwrite get function with 

```
    public function get($id, $default = NULL)
    {
        $return = FALSE;
        if (Kohana::$profiling === TRUE and class_exists('Profiler', FALSE)) {
            // Start a new benchmark
            $benchmark = Profiler::start('Cache get', __FUNCTION__ . ' ' . $id);
        }
        if (Kohana::$caching) {
            try {
                $return = parent::get($id, $default);
            } catch (RuntimeException $e) {
                $return = false;
            }
        }
        if (isset($benchmark)) {
            // Stop the benchmark
            Profiler::stop($benchmark);
        }
        return $return;
    }
```

**Comment by acoulton** [comment]
If you want to submit a PR for catching the RuntimeException in Kohana_Cache_File I think that would be a good solution.

Although it does still sound to me like your site might benefit from using a more robust/performant cache backend. Kohana_Cache_File is simple, but slow and not particularly designed for concurrency.

**Comment by farazmunj** [comment]
Yes more robust cache might fix this issue, and I am already using Memcached for storing more frequently used object. I am using File cache to store rendered HTML/js/css segments just like Kohana Fragments so it will take more space in memory using Memcached.

**Comment by farazmunj** [comment]
Hi Acoulton,
Catching RuntimeException exception is working good. can you guide me how to submit a PR do I make changes on 3.3 master branch or or 3,3 Dev?

**Comment by acoulton** [comment]
Great - please target it to 3.3/develop. Ideally we'd ask for a unit test too, but I think this will probably be quite hard to reproduce in test so don't worry if you can't.
